Basic codec setup

surma opened the pull request:
Closes #6

surma commented:
Eugh, forgot description.

Please take a look if you think the setup — described in README — is useful this way.

I am hoping we can implement a common interface for all codecs, time will tell :)

jakearchibald commented on codecs/README.md:
````diff
@@ -0,0 +1,32 @@
+$ npm run build
+```
+
+This will build two files: `<codec name>_codec.js` and `<codec name>_codec.wasm`. Due to some current limitations in Emscripten, both of these files have to be served from the root folder. When the `.js` file is loaded, a global `<codec name>` is created with the same API as an [Emscripten `Module`](https://kripken.github.io/emscripten-site/docs/api_reference/module.html).
````
"folder"? As in, the root of the server? Is there an emscripten ticket to remove this limitation?

jakearchibald replied:
Seems a shame that it creates a global. Is there an emscripten ticket for creating modules?

surma replied:
Not sure, I mentioned it to the wasm people on the feasibility study, but haven’t followed up. Basically, the path that you use for output is hard-coded in the the JS file >.>

jakearchibald replied:
Seems like `Module["locateFile"]` is the answer https://kripken.github.io/emscripten-site/docs/api_reference/module.html#affecting-execution

surma replied:
Cool. Missed that. I’ll remove the remark.

jakearchibald commented on codecs/README.md:
```diff
@@ -0,0 +1,32 @@
+
+Each codec offers at least 2 functions:
+
+- `void encode(uint8_t* img_in, int width, int height, float quality)`
```
We're probably going to end up with very different calls for each codec, but I guess in, width, & height will be consistent.

surma replied:
I thought we could keep it consistent across all codecs. I guess the options are gonna make that impossible?

jakearchibald replied:
Yeah, I think the options make it impossible. Even `quality` doesn't make sense for lossless codecs.

Keeping the first 3 args in, width & height seems sensible though.

surma replied:
At that point we’ll have to write a JS wrapper for the module anyways. 🤷‍♂️

jakearchibald replied:
Yeah, I think we'll end up with something like that anyway to wrap up an array of output bytes along with a `.free()` method.

jakearchibald commented on codecs/README.md:
```diff
@@ -0,0 +1,32 @@
+Each codec offers at least 2 functions:
+
+- `void encode(uint8_t* img_in, int width, int height, float quality)`
+- `void decode(???)`
```
Maybe add a TODO note?

jakearchibald replied:
We should avoid bundling the decoder into the same file though, since all browsers have a JPEG decoder.

surma replied:
LOL I left a comment and forgot to submit it >.>

Good call regarding splitting encoder and decoder. I’ll adjust wording.

jakearchibald commented on codecs/mozjpeg/mozjpeg.c:
```diff
@@ -0,0 +1,160 @@
+
+EMSCRIPTEN_KEEPALIVE
+int version() {
+  return 0x030301; // Lol harcoded
```
Why?

surma replied:
I honestly don’t remember. Let me check.

jakearchibald commented on codecs/mozjpeg/mozjpeg.c:
```diff
@@ -0,0 +1,160 @@
+
+EMSCRIPTEN_KEEPALIVE
+void free_result(uint8_t* result) {
+  free(result); // not sure if this is right with mozjpeg
```
Open a bug for us to look into this?

surma commented:
> I usually avoid checking in built files

Yup. I fucked up. Will fix.

surma commented:
@jakearchibald PTAL

I added per-codec documentation and example files.

I also fixed the hard-coded MozJPEG version number. Was quite the journey, but probably worth figuring out correctly.

surma commented on codecs/webp_enc/example.html:
```diff
@@ -0,0 +1,47 @@
+    await new Promise(resolve => img.onload = resolve);
+    // Make canvas same size as image
+    const canvas = document.createElement('canvas');
+    [canvas.width, canvas.height] = [img.width, img.height];
```
I mostly like this pattern to have “setting the size” as one step. Can change if you want.

jakearchibald replied:
I'd want it changed if it was in the app code rather than an example, but don't really mind here.

jakearchibald commented on codecs/mozjpeg_enc/README.md:
```diff
@@ -0,0 +1,38 @@
+
+## API
+
+### `int version()`
```
Ah, I misread the code. My bad. Yep, number is best.

jakearchibald commented on codecs/webp_enc/README.md:
```diff
@@ -0,0 +1,38 @@
+
+## API
+
+### `int version()`
```
Yeah, I misread. Ignore me!

jakearchibald commented:
Only outstanding issue is the two `example.png`s.

surma commented:
I doubled the image because then you could just spawn a webserver in each of the folders, but actually I think I can make it work with a root server or something. Will fix after meeting :)

surma commented:
Deduplicated. 

@jakearchibald PTAL
